File: implementation/scipamato/src/main/java/ch/difty/scipamato/web/jasper/summary/PaperSummaryDataSource.java
package ch.difty.scipamato.web.jasper.summary;

import java.util.Arrays;
import java.util.Collection;

import ch.difty.scipamato.entity.Paper;
import ch.difty.scipamato.entity.filter.PaperSlimFilter;
import ch.difty.scipamato.entity.projection.PaperSlim;
import ch.difty.scipamato.lib.AssertAs;
import ch.difty.scipamato.service.PaperService;
import ch.difty.scipamato.web.jasper.JasperPaperDataSource;
import ch.difty.scipamato.web.jasper.ReportHeaderFields;
import ch.difty.scipamato.web.jasper.ScipamatoPdfResourceHandler;
import ch.difty.scipamato.web.pages.paper.provider.AbstractPaperSlimProvider;
import ch.difty.scipamato.web.resources.jasper.PaperSummaryReportResourceReference;
import net.sf.jasperreports.engine.JasperReport;
import net.sf.jasperreports.export.PdfExporterConfiguration;

/**
 * DataSource for the PaperSummaryReport.
 *
 * Can be instantiated in different ways, either by passing in
 *
 * <ul>
 * <li> a single {@link Paper} + report header fields + export config</li>
 * <li> a single {@link PaperSummary} + export config</li>
 * <li> a collection of {@link PaperSummary} entities + export config or</li>
 * <li> an instance of a {@link AbstractPaperSlimProvider} + report header fields + export config</li>
 * </ul>
 *
 * The report header fields are not contained within a paper instance and make up e.g. localized labels, the brand or part of the header.
 *
 * @author u.joss
 */
public class PaperSummaryDataSource extends JasperPaperDataSource<PaperSummary> {

    private static final long serialVersionUID = 1L;

    private static final String BASE_NAME_SINGLE = "paper_summary_no_";
    private static final String BASE_NAME_SINGLE_FALLBACK = "paper_summary";
    private static final String BASE_NAME_MULTIPLE = "paper_summaries";

    private ReportHeaderFields reportHeaderFields;

    /**
     * Build up the paper summary from a {@link Paper} and any additional information not contained within the paper
     * @param paper
     *      an instance of {@link Paper} - must not be null.
     * @param reportHeaderFields
     *      collection of localized labels for report fields
     * @param config {@link PdfExporterConfiguration}
     */
    public PaperSummaryDataSource(final Paper paper, final ReportHeaderFields reportHeaderFields, PdfExporterConfiguration config) {
        this(Arrays.asList(new PaperSummary(AssertAs.notNull(paper, "paper"), AssertAs.notNull(reportHeaderFields, "reportHeaderFields"))), config,
                makeSinglePaperBaseName(paper.getNumber() != null ? String.valueOf(paper.getNumber()) : null));
        this.reportHeaderFields = reportHeaderFields;
    }

    /**
     * Populate the report from a single {@link PaperSummary}, using a specific file name including the id of the paper.
     * @param paperSummary
     *     collection of {@link PaperSummary} instances - must not be null
     * @param config
     *     the {@link PdfExporterConfiguration}
     */
    public PaperSummaryDataSource(final PaperSummary paperSummary, PdfExporterConfiguration config) {
        this(Arrays.asList(AssertAs.notNull(paperSummary, "paperSummary")), config, makeSinglePaperBaseName(paperSummary.getNumber()));
    }

    private PaperSummaryDataSource(final Collection<PaperSummary> paperSummaries, PdfExporterConfiguration config, String baseName) {
        super(new ScipamatoPdfResourceHandler(config), baseName, paperSummaries);
    }

    /**
     * Using the dataProvider for the Result Panel as record source. Needs the {@link PaperService} to retrieve the papers
     * based on the ids of the {@link PaperSlim}s that are used in the dataProvider.
     * @param dataProvider
     *      the {@link AbstractPaperSlimProvider} - must not be null
     * @param reportHeaderFields
     *      collection of localized labels for the report fields
     * @param config
     *      {@link PdfExporterConfiguration}
     */
    public PaperSummaryDataSource(final AbstractPaperSlimProvider<? extends PaperSlimFilter> dataProvider, final ReportHeaderFields reportHeaderFields, PdfExporterConfiguration config) {
        super(new ScipamatoPdfResourceHandler(config), BASE_NAME_MULTIPLE, dataProvider);
        this.reportHeaderFields = reportHeaderFields;
    }

    @Override
    protected JasperReport getReport() {
        return PaperSummaryReportResourceReference.get().getReport();
    }

    @Override
    protected PaperSummary makeEntity(Paper p) {
        return new PaperSummary(p, reportHeaderFields);
    }

    private static String makeSinglePaperBaseName(String number) {
        if (number != null && !number.isEmpty()) {
            return BASE_NAME_SINGLE + number;
        } else {
            return BASE_NAME_SINGLE_FALLBACK;
        }
    }

}

File: core/core-web/src/main/java/ch/difty/scipamato/core/web/paper/jasper/summary/PaperSummaryDataSource.kt
package ch.difty.scipamato.core.web.paper.jasper.summary

import ch.difty.scipamato.core.entity.Paper
import ch.difty.scipamato.core.entity.PaperSlimFilter
import ch.difty.scipamato.core.web.paper.AbstractPaperSlimProvider
import ch.difty.scipamato.core.web.paper.jasper.ClusterablePdfExporterConfiguration
import ch.difty.scipamato.core.web.paper.jasper.CoreShortFieldConcatenator
import ch.difty.scipamato.core.web.paper.jasper.JasperPaperDataSource
import ch.difty.scipamato.core.web.paper.jasper.ReportHeaderFields
import ch.difty.scipamato.core.web.paper.jasper.ScipamatoPdfResourceHandler
import ch.difty.scipamato.core.web.resources.jasper.PaperSummaryReportResourceReference
import net.sf.jasperreports.engine.JasperReport

/**
 * DataSource for the PaperSummaryReport.
 *
 * Can be instantiated in different ways, either by passing in
 *
 *  * a single [Paper] + report header fields + export config
 *  * a single [PaperSummary] + export config
 *  * a collection of [PaperSummary] entities + export config or
 *  * an instance of a [AbstractPaperSlimProvider] + report header fields export config
 *
 * The report header fields are not contained within a paper instance and make
 * up e.g. localized labels, the brand or part of the header.
 */
class PaperSummaryDataSource : JasperPaperDataSource<PaperSummary> {

    private lateinit var reportHeaderFields: ReportHeaderFields
    private lateinit var shortFieldConcatenator: CoreShortFieldConcatenator

    /**
     * Build up the paper summary from a [Paper] and any additional
     * information not contained within the paper
     *
     * @param paper instance of [Paper]
     * @param reportHeaderFields collection of localized labels for report fields
     * @param shortFieldConcatenator utility bean to manage the content of the fields method/population/result
     * @param config [ClusterablePdfExporterConfiguration]
     */
    constructor(
        paper: Paper,
        reportHeaderFields: ReportHeaderFields,
        shortFieldConcatenator: CoreShortFieldConcatenator,
        config: ClusterablePdfExporterConfiguration?,
    ) : this(listOf(PaperSummary(paper, shortFieldConcatenator, reportHeaderFields)), config,
        makeSinglePaperBaseName(if (paper.number != null) paper.number.toString() else null)) {
        this.reportHeaderFields = reportHeaderFields
        this.shortFieldConcatenator = shortFieldConcatenator
    }

    /**
     * Populate the report from a single [PaperSummary], using a specific file
     * name including the id of the paper.
     *
     * @param paperSummary single [PaperSummary]
     * @param config the [ClusterablePdfExporterConfiguration]
     */
    internal constructor(
        paperSummary: PaperSummary,
        config: ClusterablePdfExporterConfiguration?,
    ) : this(listOf(paperSummary), config, makeSinglePaperBaseName(paperSummary.number))

    /**
     * Populate the report from a collection of [PaperSummaries][PaperSummary],
     * using a specific file name including the id of the paper.
     *
     * @param paperSummaries collection of [PaperSummary] instances
     * @param config the [ClusterablePdfExporterConfiguration]
     * @param baseName the file name without the extension (.pdf)
     */
    private constructor(
        paperSummaries: Collection<PaperSummary>,
        config: ClusterablePdfExporterConfiguration?,
        baseName: String,
    ) : super(ScipamatoPdfResourceHandler(config), baseName, paperSummaries)

    /**
     * Using the dataProvider for the Result Panel as record source.
     *
     * @param dataProvider the [AbstractPaperSlimProvider]
     * @param reportHeaderFields collection of localized labels for the report fields
     * @param shortFieldConcatenator utility bean to manage the content of the fields method/population/result
     * @param config [ClusterablePdfExporterConfiguration]
     */
    constructor(
        dataProvider: AbstractPaperSlimProvider<out PaperSlimFilter?>,
        reportHeaderFields: ReportHeaderFields,
        shortFieldConcatenator: CoreShortFieldConcatenator,
        config: ClusterablePdfExporterConfiguration?,
    ) : super(ScipamatoPdfResourceHandler(config), BASE_NAME_MULTIPLE, dataProvider) {
        this.reportHeaderFields = reportHeaderFields
        this.shortFieldConcatenator = shortFieldConcatenator
    }

    override fun getReport(): JasperReport = PaperSummaryReportResourceReference.get().getReport()
    override fun makeEntity(p: Paper): PaperSummary = PaperSummary(p, shortFieldConcatenator, reportHeaderFields)

    companion object {
        private const val serialVersionUID = 1L
        private const val BASE_NAME_SINGLE = "paper_summary_no_"
        private const val BASE_NAME_SINGLE_FALLBACK = "paper_summary"
        private const val BASE_NAME_MULTIPLE = "paper_summaries"

        private fun makeSinglePaperBaseName(number: String?): String {
            return if (!number.isNullOrEmpty())
                BASE_NAME_SINGLE + number
            else
                BASE_NAME_SINGLE_FALLBACK
        }
    }
}
